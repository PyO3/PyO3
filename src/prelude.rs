//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{FromPyObject, IntoPyObject};
#[allow(deprecated)]
pub use crate::conversion::{IntoPy, ToPyObject};
pub use crate::err::{PyErr, PyResult};
pub use crate::instance::{Borrowed, Bound, Py, PyObject};
pub use crate::marker::Python;
pub use crate::pycell::{PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::types::{PyAny, PyModule};

#[cfg(feature = "macros")]
pub use pyo3_macros::{
    pyclass, pyfunction, pymethods, pymodule, FromPyObject, IntoPyObject, IntoPyObjectRef,
};

#[cfg(feature = "macros")]
pub use crate::wrap_pyfunction;
#[cfg(feature = "macros")]
#[allow(deprecated)]
pub use crate::wrap_pyfunction_bound;

pub use crate::types::any::PyAnyMethods;
pub use crate::types::weakref::PyWeakrefMethods;
