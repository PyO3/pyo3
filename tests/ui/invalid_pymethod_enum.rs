use pyo3::prelude::*;

#[pyclass]
enum ComplexEnum {
    Int { int: i32 },
    Str { string: String },
}

#[pymethods]
impl ComplexEnum {
    fn mutate_in_place(&mut self) {
        *self = match self {
            ComplexEnum::Int { int } => ComplexEnum::Str { string: int.to_string() },
            ComplexEnum::Str { string } => ComplexEnum::Int { int: string.len() as i32 },
        }
    }
}

#[pyclass]
enum TupleEnum {
    Int(i32),
    Str(String),
}

#[pymethods]
impl TupleEnum {
    fn mutate_in_place(&mut self) {
        *self = match self {
            TupleEnum::Int(int) => TupleEnum::Str(int.to_string()),
            TupleEnum::Str(string) => TupleEnum::Int(string.len() as i32),
        }
    }
}

fn main() {}
