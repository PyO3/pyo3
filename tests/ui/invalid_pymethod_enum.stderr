error[E0271]: type mismatch resolving `<ComplexEnum as PyClass>::Frozen == False`
  --> tests/ui/invalid_pymethod_enum.rs:11:24
   |
11 |     fn mutate_in_place(&mut self) {
   |                        ^ expected `False`, found `True`
   |
note: required by a bound in `extract_pyclass_ref_mut`
  --> src/impl_/extract_argument.rs
   |
   | pub fn extract_pyclass_ref_mut<'a, 'holder, 'py: 'a, T: PyClass<Frozen = False>>(
   |                                                                 ^^^^^^^^^^^^^^ required by this bound in `extract_pyclass_ref_mut`

error[E0271]: type mismatch resolving `<ComplexEnum as PyClass>::Frozen == False`
 --> tests/ui/invalid_pymethod_enum.rs:9:1
  |
9 | #[pymethods]
  | ^^^^^^^^^^^^ expected `False`, found `True`
  |
note: required by a bound in `PyRefMut`
 --> src/pycell.rs
  |
  | pub struct PyRefMut<'a, 'py, T: PyClass<Frozen = False>> {
  |                                         ^^^^^^^^^^^^^^ required by this bound in `PyRefMut`
  = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<TupleEnum as PyClass>::Frozen == False`
  --> tests/ui/invalid_pymethod_enum.rs:27:24
   |
27 |     fn mutate_in_place(&mut self) {
   |                        ^ expected `False`, found `True`
   |
note: required by a bound in `extract_pyclass_ref_mut`
  --> src/impl_/extract_argument.rs
   |
   | pub fn extract_pyclass_ref_mut<'a, 'holder, 'py: 'a, T: PyClass<Frozen = False>>(
   |                                                                 ^^^^^^^^^^^^^^ required by this bound in `extract_pyclass_ref_mut`

error[E0271]: type mismatch resolving `<TupleEnum as PyClass>::Frozen == False`
  --> tests/ui/invalid_pymethod_enum.rs:25:1
   |
25 | #[pymethods]
   | ^^^^^^^^^^^^ expected `False`, found `True`
   |
note: required by a bound in `PyRefMut`
  --> src/pycell.rs
   |
   | pub struct PyRefMut<'a, 'py, T: PyClass<Frozen = False>> {
   |                                         ^^^^^^^^^^^^^^ required by this bound in `PyRefMut`
   = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)
